**Modelo Naibe Bayes**
Analisis de la base de datos Longitudinal Survey usando el modelo Naive Bayes.
0. Carga de las librerias de trabajo.

In [1]:
import pandas as pd
import numpy as np
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

1. Carga de la base de datos

In [3]:
nxl='/content/1. BD1_Longitudinal Survey_Int.xlsx'
XDB=pd.read_excel(nxl, sheet_name=0)
XDB.dropna(inplace=True) #Eliminar celdas vacias

XDB=XDB[['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income', 'Urban']]
#Datos de Entrada
XD=np.array(XDB[['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income']])
#Datos de Salida - Valor deseado - Lo que quiero que aprenda el modelo
yd=np.array(XDB['Urban'])
XDB

Age  Mother_Edu  Father_Edu  FamilySize  Self_Esteem  Height  Weight  \
1       20         5.0         8.0           5         16.0    62.0   120.0   
5       19        12.0        12.0           4         27.0    64.0   200.0   
6       15        12.0        12.0           3         26.0    65.0   131.0   
7       21         9.0         6.0           3         23.0    65.0   179.0   
8       16        12.0        10.0           6         26.0    66.0   145.0   
...    ...         ...         ...         ...          ...     ...     ...   
12557   20        12.0        12.0           5         20.0    60.0   110.0   
12565   20        12.0        11.0           2         24.0    63.0   152.0   
12647   21        16.0        12.0           1         30.0    70.0   180.0   
12651   21        16.0        16.0           1         27.0    73.0   170.0   
12678   21         8.0         8.0           1         19.0    70.0   180.0   

        Income  Urban  
1          0.0    1.0  
5      40000.0    1.0  
6      25000.0    1.0  
7      27400.0    1.0  
8      52000.0    1.0  
...        ...    ...  
12557   1000.0    1.0  
12565  38000.0    1.0  
12647  43600.0    1.0  
12651  40000.0    1.0  
12678  17000.0    0.0  

[5683 rows x 9 columns]

2. Implemetar el modelo

In [5]:
mnb=GaussianNB()
mnb.fit(XD,yd)
ydp=mnb.predict(XD)

#Parametros del modelo por categoria - Urbana o Rural
u=mnb.theta_
sigma=np.sqrt(mnb.var_)

#Metrica de evaluación - Como se comporto el modelo identificando Urbano o Rural
cm=confusion_matrix(yd,ydp)
print("La matriz de confusion es:")
print(cm)

La matriz de confusion es:
[[   2 1171]
 [   4 4506]]


3. Evaluacion del modelo

In [7]:
VN=cm[0,0] #Verdaderos Negativos
FP=cm[0,1] #Falsos Positivos
FN=cm[1,0] #Falsos Negativos
VP=cm[1,1] #Verdaderos Positivos

Exactitud=(VP+VN)/len(XD)
print("La exactitud del modelo es:", Exactitud) #Comportamiento General

TE=(FN+FP)/len(XD)
print("La tasa de error del modelo es:", TE) #Porcentaje de Equivocaciones

Sensibilidad=VP/(VP+FP)
print("La sensibilidad del modelo es:", Sensibilidad) #Como se comporto frente a los positivos

Especificidad=VN/(VN+FN)
print("La especificidad del modelo es:", Especificidad) #Como se comporto frente a los negativos

Precision=VP/(VP+FN)
print("La precición del modelo es:", Precision) #Como le pego a los clasificados positivos


La exactitud del modelo es: 0.7932430054548654
La tasa de error del modelo es: 0.2067569945451346
La sensibilidad del modelo es: 0.793729082261758
La especificidad del modelo es: 0.3333333333333333
La precición del modelo es: 0.9991130820399113


4. Ejemplo de una persona especifica con caracteristicas dadas

In [104]:
Xun=np.array([[34,9,6,3,30,65,135,50000]]) #Persona especifica
ydun=mnb.predict(Xun)
print(ydun)

[1.]


Segun el modelo esta persona en especifico seria perteneciente al grupo 1 el cual es Urbano

5. Porcentje de los datos de las 2 variables en Urban

In [108]:
XDB['Urban'].value_counts()
U=(4510/5683)*100
R=(1173/5683)*100
print("El porcentaje de los datos de las 2 variables en Urban es:")
print("El porcentaje de datos para Urbano es:",U)
print("El porcentaje de datos para Rural es:",R)

El porcentaje de los datos de las 2 variables en Urban es:
El porcentaje de datos para Urbano es: 79.35949322540911
El porcentaje de datos para Rural es: 20.640506774590886


Analisis de Resultados:

En cuanto a la cantidad de resultados se puede decir que en un inicio habia 12.687 datos y despues de quitar los datos vacios teerminamos con 5.683 datos.

En cuanto a los porcentajes de los datos se descubrio que el 79,35% de los datos pertenece al grupo de Urbano y el 20,64 pertenece a rural.

El intervalo
De los resultados arrojados por el modelo se puede observar que con respecto a las metricas el modelo funicono mejor clasificacndo los del grupo de Urbano, como lo dice el indicador de sensibilidad el cual alcanzo el 79% de certeza.

De manera general se puede observar que el modelo logro una exactitud que se ubico por encima del 79% en promedio, y una tasa de error cercana al 20% en promedio.De manera general esta exactitud lo convierte en un buen modelo.

Cabe aclarar que en la variable urban el 0 corresponde a rural y el 1 corresponde a urbano y con respecto al personaje especifico se puede decir que pertenece al grupo 1 el cual es urbano segun la informacion brindada por el modelo. Y ademas se puede decir que el 79.35% de los datos son pertenecientes a urbano.

6. Porcentaje de los datos por categoria

In [8]:
Age_porcentaje = XDB['Age'].value_counts(normalize=True) * 100
print("El porcentaje de los datos en edad es:",Age_porcentaje)

MotherEdu_porcentaje = XDB['Mother_Edu'].value_counts(normalize=True) * 100
print("El porcentaje de los datos en la educacion del padre es:",MotherEdu_porcentaje)

FatherEdu_porcentaje = XDB['Father_Edu'].value_counts(normalize=True) * 100
print("El porcentaje de los datos en la educacion del padre es:",FatherEdu_porcentaje)

Height_porcentaje = XDB['Height'].value_counts(normalize=True) * 100
print("El porcentaje de los datos en la altura es:",Height_porcentaje)

Weight_porcentaje = XDB['Weight'].value_counts(normalize=True) * 100
print("El porcentaje de los datos en el peso es:",Weight_porcentaje)

Income_porcentaje = XDB['Income'].value_counts(normalize=True) * 100
print("El porcentaje de los datos en los ingresos es:",Income_porcentaje)

SelfEsteem_porcentaje = XDB['Self_Esteem'].value_counts(normalize=True) * 100
print("El porcentaje de los datos en la autoestima es:",SelfEsteem_porcentaje)

FamilySize_porcentaje = XDB['FamilySize'].value_counts(normalize=True) * 100
print("El porcentaje de los datos en el tamaño de la familia es:",FamilySize_porcentaje)

Urban_porcentaje = XDB['Urban'].value_counts(normalize=True) * 100
print("El porcentaje de los datos en la localidad es:",Urban_porcentaje)

El porcentaje de los datos en edad es: Age
17    14.235439
18    13.936301
19    13.672356
16    13.584374
15    11.543199
20    11.261658
21    11.032905
22    10.733767
Name: proportion, dtype: float64
El porcentaje de los datos en la educacion del padre es: Mother_Edu
12.0    42.688721
11.0     8.710188
10.0     7.865564
16.0     6.405068
8.0      5.666021
14.0     5.190920
9.0      4.821397
13.0     3.519268
6.0      3.360901
7.0      1.794827
15.0     1.513285
5.0      1.513285
4.0      1.425304
3.0      1.372515
17.0     1.073377
0.0      1.073377
18.0     0.791835
2.0      0.721450
1.0      0.193560
19.0     0.175963
20.0     0.123174
Name: proportion, dtype: float64
El porcentaje de los datos en la educacion del padre es: Father_Edu
12.0    34.594404
16.0     8.534225
8.0      7.636812
10.0     6.686609
14.0     5.507654
11.0     5.454865
9.0      5.226113
6.0      4.451874
13.0     2.850607
7.0      2.709836
18.0     2.164350
4.0      1.918001
0.0      1.812423
3.0      1.7948

7. Intervalo de variacion de la edad

In [19]:
min_age = XDB['Age'].min()
max_age = XDB['Age'].max()
print("El intervalo de la variable Edad es: [{}, {}]".format(min_age, max_age))

print("La media de la edad es de aproximadamente:",(22+15)/2)

El intervalo de la variable Edad es: [15, 22]
La media de la edad es de aproximadamente: 18.5


Con estos valores podemos decir que la variacion de la edad es de aproximadamente 3.5 para ambos extremos de los datos, teniendo aproximadamente 18.5 en la media.

8. Intervalo de variacion de la altura

In [18]:
min_height = XDB['Height'].min()
max_height = XDB['Height'].max()
print("El intervalo de la variable Height es: [{}, {}]".format(min_height, max_height))

print("La media de la altura es de aproximadamente:",(51+80)/2)

El intervalo de la variable Height es: [51.0, 80.0]
La media de la altura es de aproximadamente: 65.5


Con estos valores podemos decir que la variacion de la altura es de aproximadamente 14.5 para ambos extremos de los datos, teniendo aproximadamente 65.5 en la media.

9. Intervalo de variacion del peso

In [20]:
min_weight = XDB['Weight'].min()
max_weight = XDB['Weight'].max()
print("El intervalo de la variable Weight es: [{}, {}]".format(min_weight, max_weight))

print("La media del peso es de aproximadamente:",(72+375)/2)

El intervalo de la variable Weight es: [72.0, 375.0]
La media del peso es de aproximadamente: 223.5


Con estos valores podemos decir que la variacion del peso es de aproximadamente 151.5 para ambos extremos de los datos, teniendo aproximadamente 223.5 en la media.